app/src/main/java/org/mozilla/focus/locale/screen/DefaultLanguageScreenInteractor.kt

```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.locale.screen
+
+import android.app.Activity
+import mozilla.components.support.locale.LocaleManager
+import mozilla.components.support.locale.LocaleUseCases
+import org.mozilla.focus.locale.Locales
+import org.mozilla.focus.locale.screen.LanguageStorage.Companion.LOCALE_SYSTEM_DEFAULT
+import org.mozilla.focus.settings.InstalledSearchEnginesSettingsFragment
+import java.util.Locale
+
+interface LanguageScreenInteractor {
+    fun handleLanguageSelected(language: Language)
+}
+
+class DefaultLanguageScreenInteractor(
+    private val activity: Activity,
+    private val languageScreenStore: LanguageScreenStore,
+    private val localeUseCase: LocaleUseCases
+) : LanguageScreenInteractor {
+    override fun handleLanguageSelected(language: Language) {
+        if (languageScreenStore.state.selectedLanguage == language) {
+            return
+        }
+        languageScreenStore.dispatch(LanguageScreenAction.Select(selectedLanguage = language))
+        LanguageStorage(context = activity).saveCurrentLanguageInSharePref(language.tag)
+        setCurrentLanguage(language.tag)
```

Comment:
A pattern, that has been really successful, is to not expose any storage to the UI at all and instead handle this in a Middleware. The UI only displays state from the store, and only dispatches actions on the store. Then a Middleware takes care of calling actions on the store (or the system APIs) whenever it sees specific actions; or initializes the state when it sees the `InitAction`. This completely avoids that you have to have such "synchronizing" code where you need to let multiple things know about the new language; which doesn't scale so well.

The ContainerMiddleware is a nice example. It's the only piece of code interacting with the storage. Other code only reads and updates state:
https://github.com/mozilla-mobile/android-components/blob/main/components/feature/containers/src/main/java/mozilla/components/feature/containers/ContainerMiddleware.kt